t/t2703-refresh-pre-commit-hook.sh

```diff
@@ -0,0 +1,129 @@
+#!/bin/sh
+
+test_description='pre-commit hook'
+
+. ./test-lib.sh
+
+test_expect_success 'Initialise StGit' '
+    stg init &&
+    stg new -m "pre-commit-patch"
+    echo "new file" >> file &&
+    git add file
+    stg refresh
+'
+
+# install hook
+HOOKDIR="$(git rev-parse --git-dir)/hooks"
+HOOK="$HOOKDIR/pre-commit"
+mkdir -p "$HOOKDIR"
+write_script "$HOOK" <<EOF
+exit 0
+EOF
+
+test_expect_success 'refresh --no-verify with succeeding hook' '
+    echo "no-verify pre-commit-hook-success" >> file &&
+    stg refresh --no-verify
+'
+
+test_expect_success 'refresh --no-verify with path limiting, succeeding hook' '
+    echo "no-verify pre-commit-hook-path-limiting-success" >> file &&
+    stg refresh file --no-verify
+'
+
+test_expect_success 'refresh with succeeding hook' '
+    echo "pre-commit-hook-success" >> file &&
+    stg refresh
+'
+
+test_expect_success 'refresh with path limiting, succeeding hook' '
+    echo "pre-commit-hook-path-limiting-success" >> file &&
+    stg refresh file
+'
+
+# now a hook that fails
+write_script "$HOOK" <<'EOF'
+exit 1
+EOF
+
+test_expect_success 'refresh --no-verify with failing hook' '
+    echo "no-verify pre-commit-hook-fail" >> file &&
+    stg refresh --no-verify
+'
+
+test_expect_success 'refresh --no-verify with path limiting, failing hook' '
+    echo "no-verify pre-commit-hook-path-limiting-fail" >> file &&
+    stg refresh file --no-verify
+'
+
+test_expect_success 'refresh with failing hook' '
+    echo "pre-commit-hook-fail" >> file &&
+    command_error stg refresh 2>&1 |
+    grep -e "pre-commit hook failed" &&
+    git restore --staged file
+'
+
+test_expect_success 'refresh with path limiting, failing hook' '
+    echo "pre-commit-hook-path-limiting-fail" >> file &&
+    command_error stg refresh file 2>&1 |
+    grep -e "pre-commit hook failed" &&
+    git restore --staged file
+'
+
+chmod -x "$HOOK"
+test_expect_success 'refresh --no-verify with non-executable hook' '
+    echo "no-verify pre-commit-hook-non-executable" >> file &&
+    stg refresh --no-verify
+'
+
+test_expect_success 'refresh with non-executable hook' '
+    echo "pre-commit-hook-non-executable" >> file &&
+    stg refresh
+'
+
+# now a hook that edits the files added in index
+# The hook tests for trailing white spaces
+# If finds files, then fixes them and returns non-zero exit status
+test_expect_success 'install pre-commit hook' '
+    cp "$TEST_DIRECTORY"/t2703/pre-commit $HOOK &&
+    chmod +x "$HOOK"
```

Comment:
This script could be boiled-down such that it could reasonably be inlined here instead of existing as a separate file.

```sh
git diff-index --check HEAD -- && exit
sed -i 's/[[:space:]]*$//' file
exit 1
```

Not a bit deal, just something to consider.